bokeh/models/graphs.py

```diff
@@ -0,0 +1,102 @@
+from .sources import ColumnDataSource
+
+from ..core.has_props import abstract
+from ..core.properties import Any, Dict, Either, Instance, Int, Seq, String
+from ..core.validation import error
+from ..core.validation.errors import MISSING_GRAPH_DATA_SOURCE_SUBCOLUMN
+from ..model import Model
+from ..util.dependencies import import_required
+
+@abstract
+class LayoutProvider(Model):
+    '''
+
+    '''
+
+    pass
+
+
+class StaticLayoutProvider(LayoutProvider):
+    '''
+
+    '''
+
+    graph_layout = Dict(Either(String, Int), Seq(Any), default={}, help="""
+    The coordinates of the graph nodes in cartesian space. The dictionary
+    keys correspond to a node index and the values are a two element sequence
+    containing the x and y coordinates of the node.
+
+    .. code-block:: python
+        {
+            0 : [0.5, 0.5],
+            1 : [1.0, 0.86],
+            2 : [0.86, 1],
+        }
+    """)
+
+
+_DEFAULT_NODE_SOURCE = lambda: ColumnDataSource(data=dict(index=[]))
+
+
+_DEFAULT_EDGE_SOURCE = lambda: ColumnDataSource(data=dict(start=[], end=[]))
+
+
+class GraphDataSource(Model):
+    '''
+
+    '''
+
+    @classmethod
+    def from_networkx(cls, G):
+        '''
+        Generate a GraphDataSource from a networkx.Graph object
+        '''
+        nx = import_required('networkx',
+                             'To use GraphDataSource.from_networkx you need network ' +
+                             '("conda install networkx" or "pip install networkx")')
+
+        nodes = G.nodes()
+        edges = G.edges()
+        edges_start = [edge[0] for edge in edges]
+        edges_end = [edge[1] for edge in edges]
+
+        node_source = ColumnDataSource(data=dict(index=nodes))
+        edge_source = ColumnDataSource(data=dict(
+            start=edges_start,
+            end=edges_end
+        ))
+
+        graph_layout = nx.circular_layout(G)
```

Comment:
there are several networkx layout methods. What's the best way to expose them all without adding a networkx dep outside of this classmethod?

maybe have users pass in callables and if func=None, then use some default:
```python
def from_networkx(cls, G, func=None):
    if func is None:
        func = nx.circular_layout
    ....
```